packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -554,212 +594,234 @@ function basicStateReducer<S>(state: S, action: BasicStateAction<S>): S {
-  workInProgressHook = createWorkInProgressHook();
+  hook.memoizedState = hook.baseState = initialState;
+  const queue = (hook.queue = {
+    last: null,
+    dispatch: null,
+    eagerReducer: reducer,
+    eagerState: initialState,
+  });
+  const dispatch: Dispatch<A> = (queue.dispatch = (dispatchAction.bind(
+    null,
+    currentlyRenderingFiber,
+    queue,
+  ): any));
+  return [hook.memoizedState, dispatch];
+}
+
+function mountReducer<S, A>(
+  reducer: (S, A) => S,
+  initialState: void | S,
+  initialAction: void | null | A,
+): [S, Dispatch<A>] {
   if (__DEV__) {
-    currentHookNameInDev = null;
+    currentHookNameInDev = 'useReducer';
   }
-  let queue: UpdateQueue<S, A> | null = (workInProgressHook.queue: any);
-  if (queue !== null) {
-    // Already have a queue, so this is an update.
-    if (numberOfReRenders > 0) {
-      // This is a re-render. Apply the new render phase updates to the previous
-      // work-in-progress hook.
-      const dispatch: Dispatch<A> = (queue.dispatch: any);
-      if (renderPhaseUpdates !== null) {
-        // Render phase updates are stored in a map of queue -> linked list
-        const firstRenderPhaseUpdate = renderPhaseUpdates.get(queue);
-        if (firstRenderPhaseUpdate !== undefined) {
-          renderPhaseUpdates.delete(queue);
-          let newState = workInProgressHook.memoizedState;
-          let update = firstRenderPhaseUpdate;
-          do {
-            // Process this render phase update. We don't have to check the
-            // priority because it will always be the same as the current
-            // render's.
-            const action = update.action;
-            if (__DEV__) {
-              enterDisallowedContextReadInDEV();
-              isInHookUserCodeInDev = true;
-            }
-            newState = reducer(newState, action);
-            if (__DEV__) {
-              exitDisallowedContextReadInDEV();
-              isInHookUserCodeInDev = false;
-            }
-            update = update.next;
-          } while (update !== null);
-
-          workInProgressHook.memoizedState = newState;
-
-          // Don't persist the state accumlated from the render phase updates to
-          // the base state unless the queue is empty.
-          // TODO: Not sure if this is the desired semantics, but it's what we
-          // do for gDSFP. I can't remember why.
-          if (workInProgressHook.baseUpdate === queue.last) {
-            workInProgressHook.baseState = newState;
-          }
+  const hook = mountWorkInProgressHook();
+  return mountReducerImpl(hook, reducer, initialState, initialAction);
+}
 
-          return [newState, dispatch];
+function updateReducerImpl<S, A>(
+  hook: Hook,
+  reducer: (S, A) => S,
+  initialState: void | S,
+  initialAction: void | null | A,
+): [S, Dispatch<A>] {
+  const queue = hook.queue;
+
+  if (queue === null) {
+    return mountReducerImpl(hook, reducer, initialState, initialAction);
+  }
```

Comment:
We can also just throw for the empty queue case since for the case where we append hooks to the end, we'll probably just want to detect that we've hit the end and then switching the dispatcher.

Reply:
The top comment is not addressed. The array tuple should be in the outer most function.

Reply:
Ugh. GitHub showed me an old commit. Looking again...